### ISSUE TITLE
Update Dockerfile to fix reported high and medium image CVEs

### ISSUE THREAD

**Issue by jimangel** [weaveworks/flux#1964]
Before I can promote a docker image to production, it need to pass a security scan with no medium or high vulns. In the case of weaveworks/flux:1.12.0 - it trips with:

name | version | CVE | level | solution
-- | -- | -- | -- | --
gnutls | 3.6.4-r0 | CVE-2019-3829 | medium | Upgrade package gnutls to version 3.6.7-r0 or above.
gnutls | 3.6.4-r0 | CVE-2019-3836 | medium | Upgrade package gnutls to version 3.6.7-r0 or above.
libbz2 | 1.0.6-r6 | CVE-2016-3189 | medium | |
libtasn1 | 4.13-r0 | CVE-2018-1000654 | high | |

The **gnutls** one is fixed in the latest ~~alpine~~ package ([version 3.6.7-r0](https://pkgs.alpinelinux.org/package/edge/main/x86/gnutls)) as of April 11th 2019.

The **libbz2** looks like it might be a false positive as this vuln was patched in [Alpine 3.4.4](https://alpinelinux.org/posts/Alpine-3.4.4-released.html) and you're currently using Alpine 3.9.3.

```
/ # fluxd
ts=2019-04-23T14:59:28.519393824Z caller=main.go:188 version=1.12.0

/ # cat /etc/os-release
NAME="Alpine Linux"
ID=alpine
VERSION_ID=3.9.3
PRETTY_NAME="Alpine Linux v3.9"
HOME_URL="https://alpinelinux.org/"
BUG_REPORT_URL="https://bugs.alpinelinux.org/"
```

**libtasn1** is an interesting one... The primary maintainer [points out](https://gitlab.com/gnutls/libtasn1/issues/4#note_96048438) that the exploit uses a developer tool (not the actual library) and is of low risk. Not sure if it's as simple as removing for now as part of the build?

Happy to help on this as I really want to promote to production - where can I find the exact Dockerfile used to build flux?

**Comment by stefanprodan** [comment]
Alpine 3.9.3 is the latest version available on Docker Hub according to https://hub.docker.com/_/alpine?tab=tags 

Here is the Dockerfile https://github.com/weaveworks/flux/blob/master/docker/Dockerfile.flux

**Comment by jimangel** [comment]
Looks like 3.9.3 was released April 8th and the package was updated on the 11th. Not sure if it would grab it on a rebuild or if it needs to be manually bumped in the Dockerfile until 3.9.4 comes out.

**Comment by stefanprodan** [comment]
That package was released on the edge channel.

**Comment by hiddeco** [comment]
Given that all mentioned CVEs only cause a disruption of service and the attacker already must have broad access to the cluster to even start utilizing them I would not include edge packages.

**Comment by jimangel** [comment]
@hiddeco I completely agree from a developer perspective. Unfortunately this means I can't leverage flux in production, it would be great if there was some common ground...

@stefanprodan feel free to close my issue if the plan is to take no action and I will drop efforts to adopt flux.

**Comment by hiddeco** [comment]
> @hiddeco I completely agree from a developer perspective. Unfortunately this means I can't leverage flux in production, it would be great if there was some common ground...

Changing to edge packages in this case turns the lights green for your security check but may create stability or other unwanted issues as they have not undergone heavy testing.

Would you be able to adapt the policy? Because in that case my advice would be to have the CVE check in place, but with the ability to make exceptions, as a lot of times the reported CVE does not form a threat to you and no action will be required (saving you time and resources).

If this isn't the case, I can provide a patched `1.12.0` with edge packages build to you, but given the above we have no intention in fixing it in `master` before the packages have been moved into the `main` channel.

**Comment by jimangel** [comment]
> Would you be able to adapt the policy?

There definitely is wiggle room to adapt the policy, but I see adapting being more of "this CVE doesn't pose a threat of <threat-level> because of reasons X, Y, and Z." Not, "this is a valid CVE, but since the patch hasn't been tested long enough, just ignore it for now." I think ultimately we'll need to make an exception for the **libtasn1** vuln once I get a little more info on it.

I don't want to create the extra work on you to create a temporary patch in edge packages, especially if there is no intention on implementing the fix in master. I can wait until Alpine cuts a new release, flux cuts a new build, and revisit this project then.

Also, I recognize the silliness of simply changing lights "green" for security, but we need those green lights to move forward or, at the very least, a logical reaction to the found CVE's validity.

**Comment by hiddeco** [comment]
> There definitely is wiggle room to adapt the policy, but I see adapting being more of "this CVE doesn't pose a threat of [threat-level] because of reasons X, Y, and Z."

I wrote down the specific reasons bellow, and it basically all comes down to us only utilizing a fraction of what `gnupg` offers.

> **[CVE-2019-3829](https://nvd.nist.gov/vuln/detail/CVE-2019-3829)**
> A memory corruption (double free) vulnerability in the certificate verification API. Any client or server application that verifies X.509 certificates with GnuTLS 3.5.8 or later is affected.

This is a dependency of `gnupg` which we install to be able to create signed git commits and tags, Flux itself does not actually use it for X.509 certificate verification so there is no attack vector and thus does not pose a threat of any level.

> **[CVE-2019-3836](https://nvd.nist.gov/vuln/detail/CVE-2019-3836)**
> It was discovered in gnutls before version 3.6.7 upstream that there is an uninitialized pointer access in gnutls versions 3.6.3 or later which can be triggered by certain post-handshake messages.

This is, again, a dependency of `gnupg` which we install to be able to create signed git commits and tags, Flux does not use any of its TLS features so there is no attack vector and thus does not pose a threat of any level.

> **[CVE-2016-3189](https://nvd.nist.gov/vuln/detail/CVE-2016-3189)**
> Use-after-free vulnerability in bzip2recover in bzip2 1.0.6 allows remote attackers to cause a denial of service (crash) via a crafted bzip2 file, related to block ends set to before the start of the block.

This is, again, a dependency of `gnupg` which we install to be able to create signed git commits and tags, `gnupg` uses it to be able to offer `bzip2` (de)compression for file encryption, which we do not make any use of (and it does thus not pose a threat of any level).

> **[CVE-2018-1000654](https://nvd.nist.gov/vuln/detail/CVE-2018-1000654)**
> [..] libtasn1-4.12 contains a DoS, specifically CPU usage will reach 100% when running asn1Paser against the POC due to an issue in _asn1_expand_object_id(p_tree), after a long time, the program will be killed. This attack appears to be exploitable via parsing a crafted file.

Libtasn1 is the ASN.1 library used in GNUTLS, as you may already guess this gets installed as a dependency of `gnutls`, which Flux does not make any use of (see CVE-2019-3829 and CVE-2019-3836, again, no threat level).

---

> Also, I recognize the silliness of simply changing lights "green" for security, but we need those green lights to move forward or, at the very least, a logical reaction to the found CVE's validity.

I hope the above will give you a solid ground to feel confident about the CVEs not forming a threat of any level.

**Comment by jimangel** [comment]
@stefanprodan @hiddeco if the attack surface is so low for what flux does but the image has bloat that triggers a ton of CVE's; is there any reason you're not compiling flux as a go binary and building from scratch?

It seems like a lot of overhead for PGP signatures.

**Comment by hiddeco** [comment]
> is there any reason you're not compiling flux as a go binary and building from scratch?

The Flux daemon is a Go binary but uses a `git` binary to perform git actions, and git uses `gnupg` for PGP actions. We are also working on validating git commit and tag signatures which will also require `gnupg` (but the mentioned CVEs will still be irrelevant when that feature lands).

**Comment by jimangel** [comment]
@hiddeco got it, thanks for the explanation!

Security still has the heebie jeebies about the CVEs, I'll need to dig more into the source code to line up your summaries with evidence.

I really like what Flux is trying to do, so if there's an opportunity to help let me know. Maybe eventually flux could leverage a golang git implementation and wrap everything up in a single binary?

**Comment by hiddeco** [comment]
> Maybe eventually flux could leverage a golang git implementation and wrap everything up in a single binary?

This is something that has been discussed within the team and would be an improvement, it does however also have some drawbacks and would have a `size/large` label attached to it (making it not likely to happen anytime soon).

One of the challenges is the choice of implementation, there are two mature options to choose from: [src-d/go-git](https://github.com/src-d/go-git) and [libgit2/git2go](https://github.com/libgit2/git2go).

The former is a Go port of Git and in my opinion not the best option, or as they put it themselves:

> git is a humongous project with years of development by thousands of contributors, making it challenging for go-git to implement all the features.
-- [src-d/go-git](https://github.com/src-d/go-git)

The latter provides bindings to [libgit2](https://libgit2.org/), a pure C implementation of Git that is being widely used in the Git industry. The problem with this library is however that the `libgit2` itself supports signing commits with `git_commit_create_with_signature`, but there is no binding available for this (yet). So we would either need to contribute to the project (which we would happily do but increases the amount of work required) or wait until it is implemented by someone else.

> I'll need to dig more into the source code to line up your summaries with evidence.

Let me know if you need any help with this, our wrapper around `git` can be found [here](https://github.com/weaveworks/flux/tree/master/git) and the wrapper around `gnupg` is [here](https://github.com/weaveworks/flux/tree/master/gpg).

**Comment by jimangel** [comment]
Thanks for the detailed info @hiddeco! I will definitely take a look to see if I can help it (looking first at how hard it would be to add a binding to libgit2).

I saw alpine [released 3.9.4](https://alpinelinux.org/posts/Alpine-3.9.4-released.html) which includes "main/gnutls: security upgrade to 3.6.7"

Is there any chance the team is willing to cut a 1.12.1 release or a "dev" release of flux that includes Alpine 3.9.4?

**Comment by hiddeco** [comment]
As our base image is `alpine:3.9` this should already be included in the latest [`weaveworks/flux-prerelease:master-` builds](https://hub.docker.com/r/weaveworks/flux-prerelease/tags).

I created a PR to ensure it, this will be included in our next patch release `1.12.3` (scheduled for May 22).

**Comment by jimangel** [comment]
Thanks for the detailed assistance and putting up with our check boxes! I'll still look into helping trim down the attack / scan surface in the future!
🥇 @stefanprodan @hiddeco
/close
